File: src/main/java/com/leijendary/spring/boot/template/mapper/DateMapper.java
package com.leijendary.spring.boot.template.mapper;

import static com.leijendary.spring.boot.template.util.RequestContext.getTimeZone;
import static java.time.Instant.ofEpochMilli;
import static java.time.OffsetDateTime.ofInstant;
import static org.mapstruct.factory.Mappers.getMapper;

import java.time.OffsetDateTime;

import org.mapstruct.Mapper;

@Mapper
public interface DateMapper {

    DateMapper INSTANCE = getMapper(DateMapper.class);

    default OffsetDateTime toOffsetDateTime(final long epochMillis) {
        final var instant = ofEpochMilli(epochMillis);
        final var zoneId = getTimeZone().toZoneId();

        return ofInstant(instant, zoneId);
    }

    default long toEpochMillis(final OffsetDateTime offsetDateTime) {
        return offsetDateTime.toInstant().toEpochMilli();
    }
}

File: src/main/kotlin/com/leijendary/spring/boot/template/core/mapper/DateMapper.kt
package com.leijendary.spring.boot.template.core.mapper

import com.leijendary.spring.boot.template.core.util.RequestContext.timeZone
import org.mapstruct.Mapper
import org.mapstruct.factory.Mappers.getMapper
import java.time.Instant.ofEpochMilli
import java.time.OffsetDateTime
import java.time.OffsetDateTime.ofInstant
import java.time.ZoneId

@Mapper
interface DateMapper {
    companion object {
        val INSTANCE: DateMapper = getMapper(DateMapper::class.java)
    }

    fun toOffsetDateTime(epochMillis: Long): OffsetDateTime {
        val instant = ofEpochMilli(epochMillis)
        val zoneId: ZoneId = timeZone.toZoneId()

        return ofInstant(instant, zoneId)
    }

    fun toEpochMillis(offsetDateTime: OffsetDateTime): Long {
        return offsetDateTime.toInstant().toEpochMilli()
    }
}